compilers/concrete-compiler/compiler/lib/Dialect/FHE/Analysis/MANP.cpp

```diff
@@ -308,759 +236,271 @@ getNoOpSqMANP(llvm::ArrayRef<const MANPLattice *> operandMANPs) {
-      // KxLxMxN @ N -> KxLxM
+  auto rhsType =
+      ((mlir::Type)op.getRhs().getType()).cast<mlir::RankedTensorType>();
 
-      for (int64_t i = 0; i < N; i++) {
-        llvm::APInt cst = denseValsAP[i];
-        llvm::APInt rhsNorm = APIntWidthExtendSqForConstant(cst);
-        llvm::APInt mulNorm = APIntWidthExtendUMul(lhsNorm, rhsNorm);
-        accNorm = APIntWidthExtendUAdd(mulNorm, accNorm);
-      }
+  llvm::ArrayRef<int64_t> rhsShape = rhsType.getShape();
 
-    } else if (rhsDims >= 2) {
+  int64_t rhsDims = (int64_t)rhsShape.size();
 
-      // KxLxMxN @   NxP -> KxLxMxP
-      // KxLxMxN @ LxNxP -> KxLxMxP
-      // Kx1xMxN @ LxNxP -> KxLxMxP
+  assert(rhsDims == 1 && "In MANP computation dot product RHS expected to have "
+                         "a single dimension");
 
-      //   MxN @ KxLxNxP -> KxLxMxP
-      // LxMxN @ KxLxNxP -> KxLxMxP
-      // 1xMxN @ KxLxNxP -> KxLxMxP
+  int64_t N = rhsShape[0];
 
-      // N @     NxP ->     P
-      // N @   LxNxP ->   LxP
-      // N @ KxLxNxP -> KxLxP
+  // Compute output MANP:
+  // Tlu output MANP is 1
+  llvm::APInt tlu = {1, 1, false};
+  // The element-wise multiplication is given by the
+  // subtraction of two TLU outputs. The MANP of the multiplication is thus
+  // the sum of the TLU MANPs
+  llvm::APInt elemMulNorm = APIntWidthExtendUAdd(tlu, tlu);
 
-      accNorm = calculateSqManpForMatMulWithDenseValues(rhsShape, rhsDims - 2,
-                                                        denseVals, lhsNorm);
-    }
+  llvm::APInt accNorm = llvm::APInt{1, 0, false};
 
-  } else {
-    llvm::APInt rhsNorm = conservativeIntNorm2Sq(rhsElementType);
-    for (int64_t i = 0; i < N; i++) {
-      llvm::APInt mulNorm = APIntWidthExtendUMul(lhsNorm, rhsNorm);
-      accNorm = APIntWidthExtendUAdd(mulNorm, accNorm);
-    }
+  // For the total Dot product MANP, take the manp of the sum of products
+  for (int64_t i = 0; i < N; i++) {
+    accNorm = APIntWidthExtendUAdd(elemMulNorm, accNorm);
   }
 
   return accNorm;
 }
 
-static llvm::APInt getSqMANP(mlir::concretelang::FHELinalg::MatMulIntEintOp op,
-                             llvm::ArrayRef<const MANPLattice *> operandMANPs) {
-
-  auto lhsType =
-      ((mlir::Type)op.getLhs().getType()).cast<mlir::RankedTensorType>();
-  auto rhsType =
-      ((mlir::Type)op.getRhs().getType()).cast<mlir::RankedTensorType>();
-
-  llvm::ArrayRef<int64_t> lhsShape = lhsType.getShape();
-  llvm::ArrayRef<int64_t> rhsShape = rhsType.getShape();
+/// Calculates the squared Minimal Arithmetic Noise Padding of an unary FHE
+/// operation.
+static std::optional<llvm::APInt>
+getSqMANP(mlir::concretelang::FHE::UnaryEint op,
+          llvm::ArrayRef<const MANPLattice *> operandMANPs) {
+  // not all unary ops taking an encrypted operand have a type signature
+  // reflecting that, a check might be required (FHELinalg.TransposeOp is one
+  // such known op)
+  if (op.operandIntType().isa<mlir::concretelang::FHE::FheIntegerInterface>()) {
```

Comment:
`TransposeOp` takes `AnyType:$tensor` as an input argument. It's not wrong because transposition doesn't look inside tensor elements, so they can be any type.